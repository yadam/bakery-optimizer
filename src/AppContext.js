import PropTypes from 'prop-types';
import React, { useState } from 'react';

export const AppContext = React.createContext();

export const AppContextProvider = ({ children }) => {
  const [breads, setBreads] = useState([]);
  const [customers, setCustomers] = useState([]);
  return (
    <AppContext.Provider value={{ breads, customers, setBreads, setCustomers }}>
      {children}
    </AppContext.Provider>
  );
};

AppContextProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node),
  ]),
};

AppContextProvider.defaultProps = {
  children: undefined,
};

export default { AppContext, AppContextProvider };
