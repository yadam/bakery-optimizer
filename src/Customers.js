import React, { memo, useContext, useState } from 'react';
import {
  Box,
  Button,
  FormControl,
  FormControlLabel,
  FormLabel,
  Grid,
  Radio,
  RadioGroup,
  Typography,
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { AppContext } from './AppContext';

const useStyles = makeStyles((theme) => ({
  root: {
    margin: theme.spacing(6, 0, 3),
  },
  input: {
    margin: theme.spacing(2, 0, 0, 0),
  },
  list: {
    padding: theme.spacing(0, 1, 1, 1),
  },
  button: {
    margin: theme.spacing(1.25, 0, 0, 1),
  },
}));

export const Customers = memo(() => {
  const { breads, customers, setCustomers } = useContext(AppContext);
  const [currentCustomer, setCurrentCustomer] = useState({});
  const classes = useStyles();

  const radioChange = (bread) => (e) => {
    setCurrentCustomer({ ...currentCustomer, [bread]: e.target.value });
  };

  const addCustomer = () => {
    if (Object.keys(currentCustomer).length) {
      const customer = {
        ...currentCustomer,
        key: customers.length + 1,
      };
      setCustomers([...customers, customer]);
      // clear radio button current state
      setCurrentCustomer({});
    }
  };

  return (
    <Box className={classes.root}>
      <Typography className={classes.list}>Customers:</Typography>
      {customers.map((customer, index) => (
        <div key={customer.key}>
          {index + 1}:&nbsp;
          {breads.map((bread) => (
            <span key={`${customer.key}-${bread}`}>
              {bread} -&gt; {customer[bread] || 'none'}&nbsp;
            </span>
          ))}
        </div>
      ))}
      <Grid container className={classes.input}>
        <Grid xs={10} item>
          {breads.map((bread) => (
            <FormControl key={`customer-radio-${bread}`} component="fieldset">
              <FormLabel component="legend">{bread}</FormLabel>
              <RadioGroup
                aria-label={bread}
                name={bread}
                value={currentCustomer[bread] || ''}
                onChange={radioChange(bread)}
              >
                <FormControlLabel
                  value="round"
                  control={<Radio />}
                  label="Round Loaf"
                />
                <FormControlLabel
                  value="pan"
                  control={<Radio />}
                  label="Pan Loaf"
                />
              </RadioGroup>
            </FormControl>
          ))}
        </Grid>
        <Grid xs={2} item>
          <Button
            variant="contained"
            color="primary"
            className={classes.button}
            fullWidth
            onClick={addCustomer}
          >
            Add
          </Button>
        </Grid>
      </Grid>
    </Box>
  );
});

export default Customers;
