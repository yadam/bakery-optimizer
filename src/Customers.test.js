import React from 'react';
import { act, create } from 'react-test-renderer';
import { AppContext } from './AppContext';
import { Customers } from './Customers';

const breads = ['sourdough', 'wheat', 'banana'];
const customers = [
  {
    key: 0,
    sourdough: 'round',
  },
  {
    key: 1,
    wheat: 'pan',
    banana: 'round',
  },
  {
    key: 2,
    sourdough: 'pan',
    wheat: 'round',
  },
];

describe('Customers', () => {
  it('renders', () => {
    const setCustomersMock = jest.fn();
    const tree = create(
      <AppContext.Provider
        value={{ breads: [], customers: [], setCustomers: setCustomersMock }}
      >
        <Customers />
      </AppContext.Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with customers', () => {
    const setCustomersMock = jest.fn();
    const tree = create(
      <AppContext.Provider
        value={{ breads, customers, setCustomers: setCustomersMock }}
      >
        <Customers />
      </AppContext.Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('adds a customer', () => {
    const setCustomersMock = jest.fn();
    const tree = create(
      <AppContext.Provider
        value={{ breads, customers, setCustomers: setCustomersMock }}
      >
        <Customers />
      </AppContext.Provider>,
    );
    const instance = tree.root;
    const input = instance.findByProps({ name: 'sourdough' });
    const button = instance.findByType('button');
    act(() => {
      input.props.onChange({ target: { value: 'sourdough' } });
    });
    act(() => {
      button.props.onClick();
    });
    expect(setCustomersMock).toHaveBeenCalledTimes(1);
  });

  it('does not add a customer if nothing selected', () => {
    const setCustomersMock = jest.fn();
    const tree = create(
      <AppContext.Provider
        value={{ breads, customers, setCustomers: setCustomersMock }}
      >
        <Customers />
      </AppContext.Provider>,
    );
    const instance = tree.root;
    const button = instance.findByType('button');
    act(() => {
      button.props.onClick();
    });
    expect(setCustomersMock).not.toHaveBeenCalled();
  });
});
