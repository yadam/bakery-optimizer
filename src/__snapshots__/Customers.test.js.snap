// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customers renders 1`] = `
<div
  className="MuiBox-root MuiBox-root-5 makeStyles-root-1"
>
  <p
    className="MuiTypography-root makeStyles-list-3 MuiTypography-body1"
  >
    Customers:
  </p>
  <div
    className="MuiGrid-root makeStyles-input-2 MuiGrid-container"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
    />
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-4 MuiButton-containedPrimary MuiButton-fullWidth"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Add
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Customers renders with customers 1`] = `
<div
  className="MuiBox-root MuiBox-root-6 makeStyles-root-1"
>
  <p
    className="MuiTypography-root makeStyles-list-3 MuiTypography-body1"
  >
    Customers:
  </p>
  <div>
    1
    : 
    <span>
      sourdough
       -&gt; 
      round
       
    </span>
    <span>
      wheat
       -&gt; 
      none
       
    </span>
    <span>
      banana
       -&gt; 
      none
       
    </span>
  </div>
  <div>
    2
    : 
    <span>
      sourdough
       -&gt; 
      none
       
    </span>
    <span>
      wheat
       -&gt; 
      pan
       
    </span>
    <span>
      banana
       -&gt; 
      round
       
    </span>
  </div>
  <div>
    3
    : 
    <span>
      sourdough
       -&gt; 
      pan
       
    </span>
    <span>
      wheat
       -&gt; 
      round
       
    </span>
    <span>
      banana
       -&gt; 
      none
       
    </span>
  </div>
  <div
    className="MuiGrid-root makeStyles-input-2 MuiGrid-container"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
    >
      <fieldset
        className="MuiFormControl-root"
      >
        <legend
          className="MuiFormLabel-root"
        >
          sourdough
        </legend>
        <div
          aria-label="sourdough"
          className="MuiFormGroup-root"
          role="radiogroup"
        >
          <label
            className="MuiFormControlLabel-root"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-7 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-10"
                  disabled={false}
                  name="sourdough"
                  onChange={[Function]}
                  type="radio"
                  value="round"
                />
                <div
                  className="PrivateRadioButtonIcon-root-11"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root PrivateRadioButtonIcon-layer-12"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </span>
            <span
              className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Round Loaf
            </span>
          </label>
          <label
            className="MuiFormControlLabel-root"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-7 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-10"
                  disabled={false}
                  name="sourdough"
                  onChange={[Function]}
                  type="radio"
                  value="pan"
                />
                <div
                  className="PrivateRadioButtonIcon-root-11"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root PrivateRadioButtonIcon-layer-12"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </span>
            <span
              className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Pan Loaf
            </span>
          </label>
        </div>
      </fieldset>
      <fieldset
        className="MuiFormControl-root"
      >
        <legend
          className="MuiFormLabel-root"
        >
          wheat
        </legend>
        <div
          aria-label="wheat"
          className="MuiFormGroup-root"
          role="radiogroup"
        >
          <label
            className="MuiFormControlLabel-root"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-7 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-10"
                  disabled={false}
                  name="wheat"
                  onChange={[Function]}
                  type="radio"
                  value="round"
                />
                <div
                  className="PrivateRadioButtonIcon-root-11"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root PrivateRadioButtonIcon-layer-12"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </span>
            <span
              className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Round Loaf
            </span>
          </label>
          <label
            className="MuiFormControlLabel-root"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-7 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-10"
                  disabled={false}
                  name="wheat"
                  onChange={[Function]}
                  type="radio"
                  value="pan"
                />
                <div
                  className="PrivateRadioButtonIcon-root-11"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root PrivateRadioButtonIcon-layer-12"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </span>
            <span
              className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Pan Loaf
            </span>
          </label>
        </div>
      </fieldset>
      <fieldset
        className="MuiFormControl-root"
      >
        <legend
          className="MuiFormLabel-root"
        >
          banana
        </legend>
        <div
          aria-label="banana"
          className="MuiFormGroup-root"
          role="radiogroup"
        >
          <label
            className="MuiFormControlLabel-root"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-7 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-10"
                  disabled={false}
                  name="banana"
                  onChange={[Function]}
                  type="radio"
                  value="round"
                />
                <div
                  className="PrivateRadioButtonIcon-root-11"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root PrivateRadioButtonIcon-layer-12"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </span>
            <span
              className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Round Loaf
            </span>
          </label>
          <label
            className="MuiFormControlLabel-root"
          >
            <span
              aria-disabled={false}
              className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-7 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={null}
            >
              <span
                className="MuiIconButton-label"
              >
                <input
                  checked={false}
                  className="PrivateSwitchBase-input-10"
                  disabled={false}
                  name="banana"
                  onChange={[Function]}
                  type="radio"
                  value="pan"
                />
                <div
                  className="PrivateRadioButtonIcon-root-11"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root PrivateRadioButtonIcon-layer-12"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </span>
            <span
              className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Pan Loaf
            </span>
          </label>
        </div>
      </fieldset>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-4 MuiButton-containedPrimary MuiButton-fullWidth"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Add
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
